### ISSUE TITLE
FIRST_SWITCHED and LAST_SWITCHED on Elasticsearch

### ISSUE THREAD

**Issue by ryu80** [ntop/ntopng#911]
Hello, 
i started to use ntopng with Elastricsearch/kibana and i would like to calculate the bandwidth mean  like this 
BYTES_SENT / (LAST_SWITCHED - FIRST_SWITCHED) 
anyway sometimes the 2 times are equal (because the flow duration is less than 1 second i presume) and is not possible to calculate the bandwidth.
On the nprobe manual i see they should display
"SysUptime (msec) of the firs/last flow pkt"
but on elasticsearch the time is displayed as seconds.
Is there any way  to use milliseconds on these time counters?
Regards

**Comment by lucaderi** [comment]
I think you can just do BYTES_SENT / (LAST_SWITCHED - FIRST_SWITCHED + 1) and fix your problem.

**Comment by ryu80** [comment]
Hi Luca,
thank you for the suggestion. 
Anyway with a flow duration less than 1 second the bandwidth calculation could be underestimated or imprecise.
it should be usefull to have the %DURATION_IN/OUT field with msec values.
Will be possible or is it only available with nprobe/ELK combination?

**Comment by simonemainardi** [comment]
Flow throughput has been added to nProbe. To get the throughput in ntopng you can use nProbe template fields
- SRC_TO_DST_MAX_THROUGHPUT
- SRC_TO_DST_MIN_THROUGHPUT
- SRC_TO_DST_AVG_THROUGHPUT
- SRC_TO_DST_AVG_THROUGHPUT
- DST_TO_SRC_MIN_THROUGHPUT
- DST_TO_SRC_AVG_THROUGHPUT

Once you have added the template fields above, ntopng will export the corresponding data to either MySQL or to ElasticSearch

**Comment by ryu80** [comment]
Great improvement !
but is it possible to use them directly with ntopng community ? because i don't understand how to add templates with ntopng. 
maybe i have to modify something in ntopng_template_elk.json?
thanks

**Comment by simonemainardi** [comment]
@ryu80  you need nProbe and specify those field in nProbe template with -T. Then those fields will be automatically propagated to ntopng

**Comment by simonemainardi** [comment]
if your aim is to just have the throughput in elasticsearch, then you can also skip ntopng and let nprobe do the insertion in ES

**Comment by shaoranrch** [comment]
Hello,

I've got a question and hope I could get help with this. How one could get the avg throughput calculated over all the flows using the throughput fields? 

I want to calculate incoming and outgoing traffic as a whole over a time span (say 60 seconds) so I can print it to a graphics tool (grafana for instance) in near real time. 

I tried something like this:

```
SELECT UNIX_TIMESTAMP(NOW()) INTO @endRange;
SELECT (@endRange-60) INTO @startRange;
SELECT tb1.time, SUM(tb1.outt)/(1024*1024*60), SUM(tb1.incc)/(1024*1024*60)
FROM(
   SELECT logged_time AS time, SRC_TO_DST_AVG_THROUGHPUT AS outt, 
   DST_TO_SRC_AVG_THROUGHPUT AS incc
   FROM netflow_v9_flows
   WHERE logged_time BETWEEN @startRange AND @endRange
   AND IPV4_SRC_ADDR BETWEEN 111111111 AND 222222222
   UNION
   SELECT logged_time AS time, DST_TO_SRC_AVG_THROUGHPUT AS outt, 
   SRC_TO_DST_AVG_THROUGHPUT AS incc
   FROM netflow_v9_flows
   WHERE logged_time BETWEEN @startRange AND @endRange
   AND IPV4_DST_ADDR BETWEEN 111111111 AND 222222222) tb1
GROUP BY tb1.time
```
But this doesn't provide accurate information as compared to the other systems that are monitoring using SNMP, I don't see any other way to group by minute flows and calculating the throughput of that span in mbps. 

My reasoning is that if I group by LAST_SWITCHED I should get a consistent view of the flows up to 60 seconds prior to that moment (our routers have an active_timeout of 60 seconds). I do the UNION as referenced to our local ips to view what's IN and OUT whether we started the connection or was started outside.  However I'm not 100% sure of my reasoning, specially by seeing the results. 

Also I can see there are flows with throughputs reported as 0, any way to change this?
